test/core/derivatives/derivatives_test.py

```diff
@@ -217,7 +229,11 @@ def test_weight_hh_l0_jac_t_mat_prod(problem, sum_batch, V=4):
     [True, False],
     ids=["save_memory=True", "save_memory=False"],
 )
-@pytest.mark.parametrize("problem", PROBLEMS_WITH_WEIGHTS, ids=IDS_WITH_WEIGHTS)
+@pytest.mark.parametrize(
+    "problem",
+    PROBLEMS_WITH_WEIGHTS + BATCH_NORM_PROBLEMS,
+    ids=IDS_WITH_WEIGHTS + BATCH_NORM_IDS,
+)
```

Comment:
Needs discussion. My intuition was predicting that this would fail for `sum_batch=False`.

Reply:
Could be related to the initialization of the `weight` parameter.

Reply:
Is this still an open issue? Should we discuss this in person?

Reply:
Yes. This must probably be discussed in person, unless I find an explanation that makes sense to me.

Reply:
I think I convinced myself why this works. My intuition was tied too much to the notion of individual gradients, but if you just consider the bare autodiff operations, it makes sense (please confirm):

Consider the mapping `input = [input[0,:], ..., input[N-1,:]] -> output = [output[0,:], ..., output[N-1,:]]` by a parameter `weight`. The mapping can be anything (including batch norm),

* `sum_batch=False` applies the transpose weight Jacobian to every slice in the output. The result is
  ```python
  [
     result[0,:] = (d output[0,:] / d weight)^T mat,
     ...
     result[N-1,:] = (d output[N-1,:] / d weight)^T mat,
  ]
  ```
  My confusion here was that I considered `result[n,:]` the individual gradient for sample `n`, which is not defined
  if `loss[n]` depends on all inputs. But of course such a Jacobian-vector product can be performed since we only care  about derivatives here, rather than the meaningfulness of the resulting operations.
* `sum_batch=True` does basically the same, but returns the sum
  ```python
  (d output[0,:] / d weight)^T mat + ... + (d output[N-1,:] / d weight)^T mat
  ```

Reply:
Thus I predict that we will run into this problem when testing the `BatchGrad` extension, because there gradients are computed per-sample in a for loop and `BatchNorm([x1, x2, ..., xN]) != [BatchNorm(x1), BatchNorm(x2), ..., BatchNorm(xN)]`

Reply:
I have difficulties following your calculations...

However, I feel like one important aspect is missing in this argumentation.
If the forward pass is done with the whole batch, all subsequent computations (especially the backward pass) depend on these batch statistics. Therefore, all computed quantities depend on these batch_statistics.

I would even argue, that individual gradients are well defined in this context.

Consequently, I expect the extensions to go just fine, as long as the forward pass is done with the whole batch, even if the backward pass is executed as a for loop.

Reply:
I think what you're saying does not conflict with my description:
* The current computation of individual gradients is a for loop (both forward and backward). Hence the `BatchGrad` tests will fail
* If we modify the test to perform the forward pass full-batch, and the backward pass per-sample in a for loop, the test will pass again. **But** I would be careful calling the result an individual gradient. It is a weight Jacobian-vector product for each the individual summand of the mini-batch loss, but every summand depends on **all** samples.

Having this discussion has already helped me a lot to better understand batch norm. Let's continue this in person next week.

Reply:
More in the notes I sent you. Closing here.